test/Microsoft.ML.Functional.Tests/ModelFiles.cs

```diff
@@ -213,6 +213,43 @@ void AssertIsGam(ITransformer trans)
             Done();
         }
 
+        public class ModelInput
+        {
+#pragma warning disable SA1401
+            public string[] CategoricalFeatures;
+            public float[] NumericalFeatures;
+#pragma warning restore SA1401
+        }
+
+        public class ModelOutput
+        {
+#pragma warning disable SA1401
+            public float[] Score;
+#pragma warning restore SA1401
+        }
+
+
+        [Fact]
+        public void LoadModelWithOptionalColumnTransform()
+        {
+            SchemaDefinition inputSchemaDefinition = SchemaDefinition.Create(typeof(ModelInput));
+            inputSchemaDefinition[nameof(ModelInput.CategoricalFeatures)].ColumnType = new VectorDataViewType(TextDataViewType.Instance, 5);
+            inputSchemaDefinition[nameof(ModelInput.NumericalFeatures)].ColumnType = new VectorDataViewType(NumberDataViewType.Single, 3);
+            var mlContext = new MLContext();
+            ITransformer trainedModel;
+            DataViewSchema dataViewSchema;
+            using (var stream = new FileStream(Path.Combine(Directory.GetCurrentDirectory(), @"..\..\..\..\test\data\backcompat\modelwithoptionalcolumntransform.zip"),
+                FileMode.Open, FileAccess.Read, FileShare.Read))
+            {
+                trainedModel = mlContext.Model.Load(stream, out dataViewSchema);
+            }
+
+            var model = mlContext.Model.CreatePredictionEngine<ModelInput, ModelOutput>(trainedModel, inputSchemaDefinition: inputSchemaDefinition);
```

Comment:
>CreatePredictionEngine [](start = 40, length = 22)

Does the model file contain an input schema?
If it does, then you can pass it to `CreatePredictionEngine` and you would not need to create the `SchemaDefinition`.

Reply:
No you will need to pass the schema because of a bug introduced by your change. [https://github.com/dotnet/machinelearning/issues/3234](https://github.com/dotnet/machinelearning/issues/3234)

System.InvalidOperationException: 'Incompatible features column type: 'Vector<Single>' vs 'Vector<Single, 38>''

---
In reply to: [284452964](https://github.com/dotnet/machinelearning/pull/3700#discussion_r284452964) [](ancestors = 284452964)